secp256k1_test.go

```diff
@@ -0,0 +1,534 @@
+		}
+		pubkey.Negate()
+
+		privkeyBig.Sub(privkeyBig, &seeds[i])
+		privkeyBig.Mod(privkeyBig, Secp256k1Order)
+		if !bytes.Equal(intTo32Bytes(privkeyBig), privkey.privkey[:]) {
+			t.Errorf("Add operation failed, %v != %v", intTo32Bytes(privkeyBig), privkey.privkey)
+		}
+		tmpPubKey, err := privkey.GenerateSchnorrPublicKey()
+		if err != nil {
+			t.Fatalf("Failed generating pubkey from %v %v", err, privkey)
+		}
+		if *pubkey != *tmpPubKey {
+			t.Errorf("tweaked pubkey %v doesn't match tweaked privkey %v, %v", pubkey, privkey, tmpPubKey)
+		}
+	}
+	if privkey.privkey != originalPrivatekey.privkey {
+		t.Errorf("resulting privkey: %v doesn't match original: %v", privkey, originalPrivatekey)
+	}
+	if privkeyBigOriginal.Cmp(privkeyBig) != 0 {
+		t.Errorf("resulting bigint privkey: %v doesn't match original: %v", privkeyBig, privkeyBigOriginal)
+	}
+
+	if *pubkey != originalPubKey {
+		t.Errorf("resulting privkey: %v doesn't match original: %v", privkey, originalPrivatekey)
+	}
+}
+
+func TestParseSchnorrPubKeyFail(t *testing.T) {
+	zeros := [65]byte{}
+	_, err := ParseSchnorrPubKey(zeros[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse 65 zeros as a pubkey %v", zeros)
+	}
+	_, err = ParseSchnorrPubKey(zeros[:33])
+	if err == nil {
+		t.Errorf("Shouldn't parse 33 zeros as a pubkey %v", zeros[:33])
+	}
+	zeros[0] = 0x04
+	_, err = ParseSchnorrPubKey(zeros[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse zeroed x and y as pubkey %v", zeros)
+	}
+	pubkey, err := fastGeneratePrivateKey(t).GenerateSchnorrPublicKey()
+	if err != nil {
+		t.Fatalf("Failed generating a random pubkey: %v", err)
+	}
+	compressed, err := pubkey.SerializeCompressed()
+	if err != nil {
+		t.Fatalf("Failed serializing a pubkey: %v %v", err, pubkey)
+	}
+	oddnessByte := compressed[0]
+	compressed[0] = 0x07
+	_, err = ParseSchnorrPubKey(compressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse a compressed key starting with 0x07 %v", compressed)
+	}
+	uncompressed, err := pubkey.SerializeUncompressed()
+	if err != nil {
+		t.Fatalf("Failed serializing a pubkey: %v %v", err, pubkey)
+	}
+	uncompressed[0] = oddnessByte
+	_, err = ParseSchnorrPubKey(uncompressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse an uncompressed key starting with 0x2/0x03 %v", uncompressed)
+	}
+	uncompressed[0] = oddnessByte + 0x04
+	_, err = ParseSchnorrPubKey(uncompressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse a weird hybrid pubkey %v", uncompressed)
+	}
+	uncompressed[0] = 0x04
+	uncompressed[64] += 0x01
+	_, err = ParseSchnorrPubKey(uncompressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse a point with invalid y coordinate %v", uncompressed)
+	}
+}
+
+func TestSchnorrPublicKey_SerializeFail(t *testing.T) {
+	pubkey := SchnorrPublicKey{}
+	_, err := pubkey.SerializeCompressed()
+	if err == nil {
+		t.Errorf("Zeroed public key isn't serializable")
+	}
+	_, err = pubkey.SerializeUncompressed()
+	if err == nil {
+		t.Errorf("Zeroed public key isn't serializable")
```

Comment:
Please specify that this is for Uncompressed and the above is for Compressed in the error message.